### PR TITLE
Add stream buffer* methods

### PR DESCRIPTION
**Pull request by regiskuckaertz**

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
hmm, in the sliding test the queue goes below capacity 🤔 I consistently get the 9 and 17 dropped off the queue and never making it to the consumer side.

**Comment by vasilmkd** [comment]
I'm still puzzled why this is the case...

**Comment by regiskuckaertz** [comment]
Will look into it in more detail this week-end ✌️

File: streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala [hunk 1]
```diff
@@ -183,6 +183,73 @@ abstract class ZStream[-R, +E, +O](
       } yield pull
     }
 
+  private final def bufferSignal[E1 >: E, O1 >: O](
```
Comment (vasilmkd) [inline streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: Sorry, @regiskuckaertz, but how does this code deal with resuming after error? I believe you have removed the mechanism for that. Doesn't this just stop after it encounters the first error?

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: I was looking at `upstream` and it cannot fail from the inside, whereas in downstream, the `pull` intercepts any error and set the flag `done` to true. After that, `pull` will always yield `Pull.End`. Is my reasoning correct? I'll add safety tests to make sure of it! But this was a test to figure out why bufferSliding is behaving differently, and it seems that downstream is pulling the end signal which blows up the test with an error of `None` as it should. But I wonder why it is pulling that end signal in the first place 🤔

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: I'm talking about the `Promise` for each element that paused the pipeline until the error has been handled downstream.

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: Ha, yes that's the thing I missed. Thanks : 🙏

**Comment by vasilmkd** [comment]
I see we haven't copied the `StreamPullSafetySpec` tests. We need them too.

**Comment by regiskuckaertz** [comment]
There was nothing wrong in the `bufferSliding`, the problem was in `takeWhile` (used in `range`) that didn't signal the end of the stream with a failure but with an empty chunk _then_ a failure.

**Comment by regiskuckaertz** [comment]
that empty chunk was making the head of the queue drop the 9. or 17.

**Comment by regiskuckaertz** [comment]
@iravid hiya - hope you're all well! can you please review when you've got some time?

**Review by iravid — APPROVED** [review]
Looks great @regiskuckaertz. Thanks for picking this up!

**Pull request merged**